core/services/ocr2/plugins/promwrapper/plugin.go

```diff
@@ -80,93 +97,251 @@ var (
 }
 
-func New(plugin types.ReportingPlugin, name string, chainType string, chainID *big.Int, config types.ReportingPluginConfig) types.ReportingPlugin {
+func (*defaultPometheusBackend) SetReportDuration(labelValues []string, duration float64) {
+	promReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPometheusBackend) SetShouldAcceptFinalizedReportDuration(labelValues []string, duration float64) {
+	promShouldAcceptFinalizedReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPometheusBackend) SetShouldTransmitAcceptedReportDuration(labelValues []string, duration float64) {
+	promShouldTransmitAcceptedReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPometheusBackend) SetCloseDuration(labelValues []string, duration float64) {
+	promClose.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPometheusBackend) SetQueryToObservationLatency(labelValues []string, latency float64) {
+	promQueryToObservationLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetObservationToReportLatency(labelValues []string, latency float64) {
+	promObservationToReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetReportToAcceptFinalizedReportLatency(labelValues []string, latency float64) {
+	promReportToAcceptFinalizedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetAcceptFinalizedReportToTransmitAcceptedReportLatency(labelValues []string, latency float64) {
+	promAcceptFinalizedReportToTransmitAcceptedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func New(
+	plugin types.ReportingPlugin,
+	name string,
+	chainType string,
+	chainID *big.Int,
+	config types.ReportingPluginConfig,
+	backend PrometheusBackend,
+) types.ReportingPlugin {
+	// Apply passed-in Prometheus backend if one is given.
+	var prometheusBackend PrometheusBackend = &defaultPometheusBackend{}
+	if backend != nil {
+		prometheusBackend = backend
+	}
+
 	return &promPlugin{
-		wrapped:      plugin,
-		name:         name,
-		chainType:    chainType,
-		chainID:      chainID,
-		oracleID:     fmt.Sprintf("%d", config.OracleID),
-		configDigest: common.Bytes2Hex(config.ConfigDigest[:]),
+		wrapped:                       plugin,
+		name:                          name,
+		chainType:                     chainType,
+		chainID:                       chainID,
+		oracleID:                      fmt.Sprintf("%d", config.OracleID),
+		configDigest:                  common.Bytes2Hex(config.ConfigDigest[:]),
+		queryEndTimes:                 make(map[types.ReportTimestamp]time.Time),
+		observationEndTimes:           make(map[types.ReportTimestamp]time.Time),
+		reportEndTimes:                make(map[types.ReportTimestamp]time.Time),
+		acceptFinalizedReportEndTimes: make(map[types.ReportTimestamp]time.Time),
```

Comment:
I'm not sure if regular `map`'s are good enough because writes to them are not threadsafe (only reads are). These might need to be `sync.Map`'s instead.

Reply:
From the libocr plugin documentation:

```
// All functions on a ReportingPlugin should be thread-safe.
```

Reply:
Dang true.